Source/Details/ASDataController.mm

```diff
@@ -281,6 +282,56 @@ - (void)_repopulateSupplementaryNodesIntoMap:(ASMutableElementMap *)map
   }
 }
 
+/**
+ * Update supplementary nodes of all kinds for sections.
+ *
+ * @param map The element map into which to apply the change.
+ * @param traitCollection The trait collection needed to initialize elements
+ * @param shouldFetchSizeRanges Whether constrained sizes should be fetched from data source
+ */
+- (void)_updateSupplementaryNodesIntoMap:(ASMutableElementMap *)map
+                         traitCollection:(ASPrimitiveTraitCollection)traitCollection
+                   shouldFetchSizeRanges:(BOOL)shouldFetchSizeRanges
+                             previousMap:(ASElementMap *)previousMap
+{
+  ASDisplayNodeAssertMainThread();
+
+  NSUInteger sectionCount = [self itemCountsFromDataSource].size();
+  BOOL canDelegate = (self.layoutDelegate != nil);
+  if (sectionCount > 0) {
+    NSIndexSet *sectionIndexes = [NSIndexSet indexSetWithIndexesInRange:NSMakeRange(0, sectionCount)];
+    ASSizeRange newSizeRange;
+    for (NSString *kind in [self supplementaryKindsInSections:sectionIndexes]) {
+      NSArray<NSIndexPath *> *indexPaths = [self _allIndexPathsForItemsOfKind:kind inSections:sectionIndexes];
+      NSMutableArray<NSIndexPath *> *indexPathsToDeleteForKind = [[NSMutableArray alloc] init];
+      NSMutableArray<NSIndexPath *> *indexPathsToInsertForKind = [[NSMutableArray alloc] init];
+      // If supplementary node does exist and size is now zero, remove it.
+      // If supplementary node doesn't exist and size is now non-zero, insert one.
+      for (NSIndexPath *indexPath in indexPaths) {
+        ASCollectionElement *previousElement = [previousMap supplementaryElementOfKind:kind atIndexPath:indexPath];
+        if (canDelegate) {
+          ASCollectionLayoutContext *context = [self.layoutDelegate layoutContextWithElements:previousMap];
+          Class<ASDataControllerLayoutDelegate> layoutDelegateClass = [self.layoutDelegate class];
+          ASCollectionLayoutState *layoutState = [layoutDelegateClass calculateLayoutWithContext:context];
+          UICollectionViewLayoutAttributes *attrs = [layoutState layoutAttributesForSupplementaryElementOfKind:kind atIndexPath:indexPath];
```

Comment:
Let's not call `calculateLayoutWithContext:` in a loop on main like this. 

These methods should only gather data and not perform expensive operations like layout. The problem I guess is that the ASCollectionLayout API is intentionally opaque about what size ranges it uses to measure its nodes.

@nguyenhuy Maybe we need new API to let ASCollectionLayoutDelegates specify what supplementary items exist during this phase, so that we can put them in the map.

Reply:
I agree that we shouldn't perform a layout here just to figure out the size range. This is a limitation of ASCollectionLayout API.

Given the limited usage of ASCollectionLayout at the moment, and the possibility that it would take some time to come up with a sensible extension for its API, I'm wondering if we should aim to merge this PR without focusing much on ASCollectionLayout support, and then create a new ticket to follow up?

Reply:
ok,  I will just leave this block as todo.